SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialLatencyTracer.h

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#pragma once
+
+#include "CoreMinimal.h"
+
+#include "SpatialConstants.h"
+#include "Containers/Map.h"
+
+#if TRACE_LIB_ACTIVE
+#include "WorkerSDK/improbable/trace.h"
+#endif
+
+#include "SpatialLatencyTracer.generated.h"
+
+DECLARE_LOG_CATEGORY_EXTERN(LogSpatialLatencyTracing, Log, All);
+
+class AActor;
+class UFunction;
+class USpatialGameInstance;
+
+using TraceKey = int32;
+
+UCLASS()
+class SPATIALGDK_API USpatialLatencyTracer : public UObject
+{
+	GENERATED_BODY()
+
+public:
+
+	//////////////////////////////////////////////////////////////////////////
+	//
+	// USpatialLatencyTracer allows for tracing of gameplay events across multiple workers, from their user
+	// instigation, to their observed results. Key timings related to these events are logged throughout
+	// the Unreal GDK networking stack.
+	// These timings are logged to Google's Stackdriver (https://cloud.google.com/stackdriver/)
+	//
+	// Setup:
+	// 1. Setup a Google project with access to Stackdriver.
+	// 2. Create and download a service-account certificate
+	// 3. Set GOOGLE_APPLICATION_CREDENTIALS to certificate path
+	// 4. Set GRPC_DEFAULT_SSL_ROOTS_FILE_PATH to your `roots.pem` gRPC path
+	//
+	// Usage:
+	// 1. Register your Google's project id with `RegisterProject`
+	// 2. Start a latency trace using `BeginLatencyTrace` tagging it against an Actor's RPC
+	// 3. During the execution of the tagged RPC either;
```

Comment:
I'm struggling to go from reading these docs to knowing what I call when. What does it mean to continue a trace and what exactly does starting a trace do and where do I call this (seemingly from a blueprint) in relation to calling the RPC. Is this called by the user? As we don't seem to be calling it.

Reply:
Fleshed out documentation more, let me know if it makes more sense.

Reply:
The specific things I found confusing after reading these docs are that I don't know what I expect to happen on each call on know when I should call them.
The BeginLatency call is not tied to a specific invocation of the RPC and I can't detect if there is an existing trace when processing an RPC. So I need to write my trace code accordingly and this may as well say what that is.

Reply:
It's tied to the next invocation of a specific RPC on a specific Actor. This allows us to maintain the existing RPC interface.

Reply:
In which case are we assuming that the rpc is called basically immediately before the RPC (which I think would be reasonable)?

Reply:
Generally, but not always. If the user wants to start the trace well before calling the RPC (due to input perhaps), that's still valid.

Reply:
In which case you have an issue if you didn't want to trace all RPCs of some type (however it seems like the continue and end calls expect this to not be the case).